spacelift/resource_run.go

```diff
@@ -80,11 +154,121 @@ func resourceRunCreate(ctx context.Context, d *schema.ResourceData, meta interfa
 	}
 
-	d.SetId(mutation.ID)
+	if waitRaw, ok := d.GetOk("wait"); ok {
+		wait := expandWaitConfiguration(waitRaw.([]interface{}))
+
+		if !wait.disabled {
+			stateConf := &retry.StateChangeConf{
+				ContinuousTargetOccurence: 1,
+				Delay:                     10 * time.Second, // TODO: Delay must be a multiple of Timeout
+				MinTimeout:                10 * time.Second,
+				Pending: []string{
+					"running",
+				},
+				Target: []string{
+					"finished",
+				},
+				Refresh: checkStackStatusFunc(ctx, client, stackID, mutation.ID),
+				Timeout: d.Timeout(schema.TimeoutCreate),
+			}
+
+			finalState, err := stateConf.WaitForStateContext(ctx)
+			if err != nil {
+				if timeoutErr, ok := internal.AsError[*retry.TimeoutError](err); ok {
+					tflog.Debug(ctx, "received retry.TimeoutError from WaitForStateContext", map[string]any{
+						"stackID":       stackID,
+						"runID":         mutation.ID,
+						"lastState":     timeoutErr.LastState,
+						"expectedState": timeoutErr.ExpectedState,
+					})
+					finalState = "__timeout__"
+				} else if err == context.DeadlineExceeded {
+					tflog.Debug(ctx, "received context.DeadlineExceeded from WaitForStateContext", map[string]any{
+						"stackID": stackID,
+						"runID":   mutation.ID,
+					})
+					finalState = "__timeout__"
+				} else {
+					return diag.Errorf("failed waiting for run %s on stack %s to finish. error(%T): %+v ", mutation.ID, stackID, err, err)
+				}
+			}
 
+			switch finalState.(string) {
+			case "__timeout__":
+				if !wait.continueOnTimeout {
+					return diag.Errorf("run %s on stack %s has timed out", mutation.ID, stackID)
+				} else {
+					tflog.Info(ctx, "run timed out but continue_on_discarded=true",
+						map[string]any{
+							"stackID": stackID,
+							"runID":   mutation.ID,
+						})
+				}
+			default:
+				if !slices.Contains[[]string](wait.continueOnState, finalState.(string)) {
+					return diag.Errorf("run %s on stack %s has ended with status %s. expected %v", mutation.ID, stackID, finalState, wait.continueOnState)
+				}
+				tflog.Debug(ctx, "run finished", map[string]any{
+					"stackID":    stackID,
+					"runID":      mutation.ID,
+					"finalState": finalState,
+				})
+			}
+		}
+	}
+
+	d.SetId(mutation.ID)
 	return nil
 }
+
+func checkStackStatusFunc(ctx context.Context, client *internal.Client, stackID string, runID string) retry.StateRefreshFunc {
+	return func() (result any, state string, err error) {
+		// instead of a resource handle we return the current state as result
+		// Makes it easier to detect which end state has been reached.
+		// Otherwise we would need another GraphQL query
+		result, finished, err := getStackRunStateByID(ctx, client, stackID, runID)
+		if err != nil {
+			return
+		}
+		state = "running"
+		if finished {
+			state = "finished"
+		}
+		return
+	}
+}
+
+func getStackRunStateByID(ctx context.Context, client *internal.Client, stackID string, runID string) (string, bool, error) {
```

Comment:
Actually, I don't think this is going to work inside Spacelift.

The tokens that Spacelift runs get are more limited in scope than regular tokens and they will not allow you to make this query. I think this will require some investigation on our end - perhaps a very limited endpoint to get run status by ID?

Reply:
I created an internal ticket to support this on the backend.